Alloy Transition and using multi-block-txn

arnav-deep opened the pull request:
Transition from Ethers to ALloy and use multi-txn-block lib for sending transactions in the same block and managing confirmations.

arnav-deep committed:
update packages

arnav-deep committed:
create contract abis + remove unrequired libs + use alloy

arnav-deep committed:
refactor from ethers to alloy + implement multi-txn-block

arnav-deep committed:
refactor: remove unused vars

arnav-deep committed:
refactor: remove unused error

arnav-deep committed:
update api_impl tests

arnav-deep committed:
fix: use packed encoding

arnav-deep committed:
fix: fix signature and decoded logs

arnav-deep committed:
refactor: use hashes kind blocks + log succeful txn submission

arnav-deep committed:
fix: use job params

arnav-deep committed:
fix: use request chain job

arnav-deep committed:
fix: use packed encoding

arnav-deep committed:
fix: slice to get address

arnav-deep committed:
fix: fix tests

arnav-deep committed:
add openssl for linux_musl target

vg-27 commented on src/api_impl.rs:
```diff
@@ -351,141 +246,98 @@ async fn export_signed_registration_message(
-            "Failed to connect to the common chain http rpc server {}: {}",
-            app_state.common_chain_http_url,
-            common_chain_http_rpc_client.unwrap_err()
+        let gateways_contract = Arc::new(GatewaysContract::new(
```
can't we create gateway contract object once and use it for all chains.

arnav-deep replied:
The `gateways_contract` object gets moved in the `call()` method below and gives the `value moved` compile error. Tried cloning and it didn't work.

vg-27 commented on src/api_impl.rs:
```diff
@@ -351,141 +246,98 @@ async fn export_signed_registration_message(
+            None,
+        )
+        .await
+        .unwrap();
```
error handling?

arnav-deep replied:
Added

vg-27 commented on src/api_impl.rs:
```diff
@@ -512,23 +364,29 @@ async fn export_signed_registration_message(
+            None,
+        )
+        .await
+        .unwrap();
```
error handling?

arnav-deep replied:
Added

vg-27 commented on src/common_chain_interaction.rs:
```diff
@@ -732,59 +743,54 @@ impl ContractsClient {
+            DynSolValue::Uint(U256::from(sign_timestamp), 256),
+        ])
+        .abi_encode();
+
+        let function_selector = &keccak256(COMMON_CHAIN_RELAY_JOB_CALL.as_bytes());
+        let mut selector = [0u8; 4];
+        selector.copy_from_slice(&function_selector[..4]);
+
+        let mut txn_data = selector.to_vec();
+        txn_data.extend(encoded_tokens[32..].to_vec());
```
isn't this auto generated using contract abi ? should avoid it doing ourself, given lib is there.

arnav-deep replied:
Done

arnav-deep committed:
Update dependencies and refactor wallet handling in the application

- Updated `alloy` dependency version from `0.6.2` to `0.6.4`.
- Updated various dependencies in `Cargo.lock`, including `syn` to `2.0.90`, `http` to `1.2.0`, and several others.
- Refactored wallet handling to use `PrivateKeySigner` instead of `String` for better type safety and clarity.
- Adjusted related code in `api_impl.rs`, `main.rs`, `model.rs`, and test utilities to accommodate the new wallet structure.
- Improved error handling and response messages in the API implementation.

arnav-deep commented on src/main.rs:
```diff
@@ -14,27 +14,25 @@ mod test_util;
 use std::sync::atomic::AtomicBool;
 use std::sync::{Arc, Mutex};
 use tokio::fs;
+use tokio::sync::RwLock as TokioRwLock;
```
Yeah, updated

arnav-deep committed:
Refactor error handling in export_signed_registration_message function

- Removed unnecessary unwrap calls and replaced them with proper error handling using `let Ok(...) = ... else` syntax.
- Improved response messages for internal server errors when creating transaction managers for request and common chains.
- Enhanced code readability and maintainability in `api_impl.rs` by streamlining error management.

vg-27 commented on src/api_impl.rs:
```diff
@@ -351,141 +246,98 @@ async fn export_signed_registration_message(
 
-    let mut request_chains_data: HashMap<u64, RequestChainData> = HashMap::new();
+    let mut request_chain_data: HashMap<u64, RequestChainData> = HashMap::new();
 
```
```suggestion
let gateways_contract = GatewaysContract::new(
        app_state.gateways_contract_addr,
        common_chain_http_provider.clone(),
    );
```

arnav-deep replied:
Done

vg-27 commented on src/api_impl.rs:
```diff
@@ -351,141 +246,98 @@ async fn export_signed_registration_message(
-            "Failed to connect to the common chain http rpc server {}: {}",
-            app_state.common_chain_http_url,
-            common_chain_http_rpc_client.unwrap_err()
+        let gateways_contract = Arc::new(GatewaysContract::new(
+            app_state.gateways_contract_addr,
+            common_chain_http_provider.clone(),
```
```suggestion
```

arnav-deep replied:
Done

vg-27 reviewed (commented):
suggested changes should solve the issue of GatewayContract object recreation. 

this [comment](https://github.com/marlinprotocol/serverless-gateway/pull/12/files/b3d7c68f9ef513245999bfb3464b96124640d22d#r1871548068)

vg-27 commented on src/api_impl.rs:
```diff
@@ -351,141 +246,98 @@ async fn export_signed_registration_message(
+                ));
+            }
+        };
+        drop(gateways_contract);
```
```suggestion

```

arnav-deep replied:
Done

arnav-deep committed:
Update dependencies and refactor wallet handling

- Updated several dependencies in `Cargo.lock`.
- Refactored wallet handling in the application to use `String` instead of `PrivateKeySigner` for better clarity and consistency.
- Enhanced error handling in `api_impl.rs` by replacing unnecessary unwraps with proper error management.
- Improved code readability and maintainability across multiple files, including `common_chain_interaction.rs`, `job_subscription_management.rs`, and `model.rs`.

arnav-deep committed:
Update gateway contract addresses in configuration file

- Changed the `gateways_contract_addr` and `gateway_jobs_contract_addr` in `oyster_serverless_gateway_config.json` to new contract addresses

arnav-deep committed:
Refactor API implementation and enhance contract interactions

- Introduced `GatewayJobsContract` to manage job-related functionalities in `api_impl.rs`.
- Updated the `export_signed_registration_message` function to utilize the new contract for job relay and reassignment, improving code clarity and maintainability.
- Replaced direct encoding of transaction data with contract method calls for better abstraction and readability.
- Added URL validation for common chain HTTP and WebSocket URLs in `main.rs` to ensure proper configuration.
- Removed obsolete constants related to job relay calls from `constant.rs`, streamlining the codebase.
- Enhanced error handling and response messages throughout the API implementation for improved robustness.

arnav-deep committed:
Refactor contract interactions and clean up API implementation

- Removed direct instantiation of `GatewayJobsContract` in favor of using method calls for job relay and reassignment, enhancing code clarity.
- Updated transaction data encoding to utilize contract method calls, improving abstraction and maintainability.
- Removed obsolete constants related to job response calls from `constant.rs`, streamlining the codebase.
- Cleaned up imports and unnecessary code in `model.rs` and `common_chain_interaction.rs` for better readability.

arnav-deep committed:
Refactor request chain data handling across multiple files

- Renamed `request_chain_data` to `request_chains_data` for consistency and clarity in `api_impl.rs`, `common_chain_interaction.rs`, `job_subscription_management.rs`, and `model.rs`.
- Updated related data structures and method calls to reflect the new naming convention, enhancing code readability and maintainability.

arnav-deep committed:
Refactor gateway contract instantiation in export_signed_registration_message function

- Moved the instantiation of the `GatewaysContract` to a single location within the function to improve code clarity and reduce redundancy.
- Removed unnecessary drop statement for `gateways_contract`, streamlining the code.

vg-27 commented on src/common_chain_interaction.rs:
```diff
@@ -1416,6 +1392,7 @@ impl LogsProvider for ContractsClient {
 mod common_chain_interaction_tests {
     use std::str::FromStr;
 
+    use alloy::dyn_abi::DynSolValue;
```
is this change needed?

arnav-deep replied:
Yeah. The module code doesn't use DynSolValue anymore because of code changes. Tests still use them, so it needs to be explicitly imported here since `super::*` doesn't have that import anymore

vg-27 commented on src/api_impl.rs:
```diff
@@ -71,24 +72,67 @@ async fn inject_mutable_config(
     }
 
-    *wallet_guard = private_key_signer;
+    let registration_events_listener_active_guard = app_state
```
why is this guard needed?

arnav-deep replied:
This guard describes the start of creating the contracts_client the first time. If this guard is not taken here, then while the contracts client is being built with the previous signer, the signer can be updated midway, and since the contracts client has not been updated yet, we would not process the update logic of the mutable config.

So this is basically to synchronise the contracts client and update of the mutable config.

vg-27 replied:
I understand that. But it needs restructuring of locks, changing their name and purposes.

arnav-deep replied:
Yeah. This was repurposed here and new usage added. Will update the name to better reflect the usage.

vg-27 commented on src/api_impl.rs:
```diff
@@ -71,24 +72,67 @@ async fn inject_mutable_config(
+            }
+        }
+
+        drop(request_chains_data);
```
does it need to be dropped explicitly?

arnav-deep replied:
No. Updated.

vg-27 commented on src/api_impl.rs:
```diff
@@ -372,16 +407,19 @@ async fn export_signed_registration_message(
         let subscription_jobs = Arc::new(RwLock::new(HashMap::new()));
 
+        if *gas_wallet_hex != *app_state.wallet.read().unwrap() {
+            return HttpResponse::BadRequest().body("Gas wallet updated!");
```
It is failing http request because of some sync issue. Can't it be made synchronous, like while creating contract client object, any update to gas key should be on hold.

arnav-deep replied:
Updated to use the signer update fn instead of throwing error.

arnav-deep committed:
refactor: use run() for the txn_manager

arnav-deep commented on src/api_impl.rs:
```diff
@@ -71,24 +72,67 @@ async fn inject_mutable_config(
         .mutable_params_injected
         .store(true, Ordering::SeqCst);
 
+    drop(registration_events_listener_active_guard);
```
Updated

arnav-deep replied:
Yeah, because this write lock is taken without involving any writes and is not being used. So if not explicitly dropped, the write lock will be shown as an unused variable. Purpose for this lock mentioned in the comment below.

arnav-deep commented on src/api_impl.rs:
```diff
@@ -71,24 +72,67 @@ async fn inject_mutable_config(
+            }
+        }
+
+        drop(request_chain_data_hashmap);
```
Not needed. Updated

arnav-deep committed:
refactor: enhance gas wallet handling in export_signed_registration_message

- Updated the gas wallet retrieval to clone the value for safer usage.
- Improved error handling when the gas wallet is updated, ensuring that the private signer for request chain transaction managers is updated accordingly.
- Removed unnecessary drop statement for `request_chains_data`, streamlining the code.

vg-27 reviewed (dismissed):
LG

arnav-deep committed:
refactor: streamline wallet handling in inject_mutable_config

- Removed unnecessary locking of registration_events_listener_active, simplifying the code.
- Enhanced clarity by ensuring contracts_client_guard is locked only once.
- Improved error handling for wallet address updates, ensuring better response management.
- Cleaned up code for better readability and maintainability.

vg-27 commented on src/api_impl.rs:
```diff
@@ -129,7 +124,7 @@ async fn inject_mutable_config(
         .mutable_params_injected
         .store(true, Ordering::SeqCst);
 
-    drop(registration_events_listener_active_guard);
```
shouldn't we drop wallet_gaurd before this, reverse ordering in drop. Also, explicit drop not required.

arnav-deep replied:
Yeah, this guard is not in use anymore.

arnav-deep committed:
refactor: simplify mutable config injection in api_impl.rs

- Removed unnecessary drop statement for contracts_client_guard, enhancing code clarity.

arnav-deep committed:
refactor: improve request chain data handling in job_subscription_management

- Simplified the cloning of request chain data by directly retrieving it from the map, enhancing code clarity.
- Updated the function call to pass a reference to the request chain data, improving performance and readability.

vg-27 reviewed (approved):
LG.
